mesonbuild/compilers/mixins/metrowerks.py

```diff
@@ -0,0 +1,224 @@
+
+    def depfile_for_object(self, objfile: str) -> T.Optional[str]:
+        # Earlier versions of these compilers do not support specifying
+        # a custom name for a depfile, and can only generate '<input_file>.d'
+        return os.path.splitext(objfile)[0] + '.' + self.get_depfile_suffix()
+
+    def get_always_args(self) -> T.List[str]:
+        return ['-gccinc']
+
+    def get_buildtype_args(self, buildtype: str) -> T.List[str]:
+        return mwcc_buildtype_args[buildtype]
+
+    def get_compile_only_args(self) -> T.List[str]:
+        return ['-c']
+
+    def get_debug_args(self, is_debug: bool) -> T.List[str]:
+        return mwcc_debug_args[is_debug]
+
+    def get_dependency_gen_args(self, outtarget: str, outfile: str) -> T.List[str]:
+        # Check comment in depfile_for_object()
+        return ['-gccdep', '-MD']
+
+    def get_depfile_suffix(self) -> str:
+        return 'd'
+
+    def get_include_args(self, path: str, is_system: bool) -> T.List[str]:
+        if not path:
+            path = '.'
+        return ['-I' + path]
+
+    def get_no_optimization_args(self) -> T.List[str]:
+        return ['-opt', 'off']
+
+    def get_no_stdinc_args(self) -> T.List[str]:
+        return ['-nostdinc']
+
+    def get_no_stdlib_link_args(self) -> T.List[str]:
+        return ['-nostdlib']
+
+    def get_optimization_args(self, optimization_level: str) -> T.List[str]:
+        return mwcc_optimization_args[optimization_level]
+
+    def get_output_args(self, target: str) -> T.List[str]:
+        return ['-o', target]
+
+    def get_pic_args(self) -> T.List[str]:
+        return ['-pic']
+
+    def get_preprocess_only_args(self) -> T.List[str]:
+        return ['-preprocess']
+
+    def get_preprocess_to_file_args(self) -> T.List[str]:
+        return ['-P']
+
+    def get_pch_use_args(self, pch_dir: str, header: str) -> T.List[str]:
+        return ['-prefix', os.path.basename(header)]
+
+    def get_pch_suffix(self) -> str:
+        return 'mch'
+    
+    def get_warn_args(self, level: str) -> T.List[str]:
+        return self.warn_args[level]
+
+    def get_werror_args(self) -> T.List[str]:
+        return ['-w', 'error']
+
+    @classmethod
+    def _unix_args_to_native(cls, args: T.List[str], info: MachineInfo) -> T.List[str]:
+        result = []
+        for i in args:
+            if i.startswith('-D'):
+                i = '-D' + i[2:]
+            if i.startswith('-I'):
+                i = '-I' + i[2:]
+            if i.startswith('-Wl,-rpath='):
+                continue
+            elif i == '--print-search-dirs':
+                continue
+            elif i.startswith('-L'):
+                continue
+            result.append(i)
+        return result
+
+    def compute_parameters_with_absolute_paths(self, parameter_list: T.List[str], build_dir: str) -> T.List[str]:
+        for idx, i in enumerate(parameter_list):
+            if i[:9] == '-I':
```

Comment:
There must be something wrong... a nine chars string cannot be equal to a two chars string!

Reply:
Whoops, that's a leftover from another compiler class I used as a base. Not sure what it's used for.

Reply:
It matches on --girdir= (vala) or /LIBPATH: (MSVC) or -include=

The purpose of this function is so that the IDE introspection `*.json` files, and specifically intro-targets.json, have a `"parameters": []` entry which specifies each member of parameter_list (the actual compiler arguments run to compile a given target) with `-I<relative-path>` converted to `-I<absolute-path>`. Mismatching here is "harmless" and results in a no-op and technically, reduced functionality.

It looks like you probably copied this from the compcert compiler, which has been broken in this fashion ever since its original implementation in #7674 -- I would accept a second patch that corrects this in the compcert mixin. :)


(Please, either way, update your metrowerks code here, to use `i[:2]`.)

Reply:
(The compcert compiler presumably copied it from some code that used MSVC style or something, updated the string, but didn't update the character count.)